### ISSUE TITLE
'Never' doesn't work with Generic class

### ISSUE THREAD

**Issue by beaudrychase** [microsoft/pyright#8432]
**Describe the bug**
I'm trying to create a class that allows you to pass in whether certain types are Optional though Generics. I was able to reproduce the issue with the code snippet below. `mypy` says the snippet is valid, but `pyright` warns that there's an error.

**Code or Screenshots**

```python
from typing import Generic, Never, TypeVar

O = TypeVar("O", Never, None)


class TestClass(Generic[O]):
    status: int | O

    def __init__(self, status: int | O):
        self.status = status


a = TestClass[None](1)
# This warns: Type "Never" cannot be assigned to type variable "O@TestClass"
b = TestClass[Never](1)
```
[mypy Playground link](https://mypy-play.net/?mypy=latest&python=3.12&gist=4234ec44dcef485b40941a6dcd478325)
[pyright Playground link](https://pyright-play.net/?code=GYJw9gtgBAhgRgYygSwgBzCALrOBnLEGBLCAUywAswATAKFEigTABtWyTkwA7PAOnhJUGbFACSWMkTgcANFACyMNGmQ8A5g3DQsATzWaU6TDgCCPPQoDiZHtOQIFAOTIA3aQoAqBsmdbIMHjevgBqMCAKIO5kMKwA%2BvpoZHR0APJQALxQPsnhIAAUAERpRS4xkVDOvGQAlKl0CKxBeDlkBADCzXh4Bbb2II4A2mkAurUAXHRQM1AEMFgArngTKDw4AD5QaamzUDRkwFDx8erIWCcFeGSswArzSytrm9uT03uz17f8D8tZc1gFssGjB-l52lgui0htV7KMCgBGeoAYhylGQrQA7hE%2BKtcmQoEVXB4QEVmDAeDwwDg4ASWsgNPYaFAsGAWb4oG4IoFZASSgABcGdbp4Ip0OBgiFQnowirwpF0IA)

**Comment by max-muoto** [comment]
As an interesting note, specifying `bound=Never|None` works here.

**Comment by erictraut** [comment]
The [typing spec](https://typing.readthedocs.io/en/latest/spec/generics.html) is not clear on whether `Never` should be allowed as a constraint for a value-constrained TypeVar. In fact, the entire topic of value-constrained type variables is woefully underspecified in the spec currently. It's on a long list of items that I would like to formally include in the typing spec. 

I think there's a good argument to be made for disallowing both `Never` and `Any` as value constraints and upper bounds. 

I'm somewhat reluctant to make a change here until we have a chance to discuss this in context of the typing spec and decide collectively on the appropriate behavior.

Given that mypy already accepts this construct, that gives me some confidence that changing pyright's behavior is less likely to cause backward compatibility issues for existing code bases. 

I'm going to think about this a bit more before making a decision on how to address it.

**Comment by erictraut** [comment]
> As an interesting note, specifying bound=Never|None works here. Unsurprisingly, NoReturn also doesn't work in the same context as above though.

That shouldn't be surprising. The type `Never | None` is equivalent to `None`. The type `Never` is equivalent to `NoReturn`.

**Comment by erictraut** [comment]
This is included in pyright 1.1.372.
